modelzoo/unet.py

```diff
@@ -0,0 +1,175 @@
+        net['upconv{}'.format(depth)] = Conv2DLayer(upscaling,
+                                        num_filters=n_filters, filter_size=2, stride=2,
+                                        W=HeNormal(gain='relu'),
+                                        nonlinearity=nonlinearity)
+
+        net['bridge{}'.format(depth)] = ConcatLayer([
+                                        net['upconv{}'.format(depth)],
+                                        net['conv{}_2'.format(depth)]],
+                                        axis=1, cropping=[None, None, 'center', 'center'])
+
+        net['_conv{}_1'.format(depth)] = Conv2DLayer(net['bridge{}'.format(depth)],
+                                        num_filters=n_filters, filter_size=3, pad='valid',
+                                        W=HeNormal(gain='relu'),
+                                        nonlinearity=nonlinearity)
+
+        net['_conv{}_2'.format(depth)] = Conv2DLayer(net['_conv{}_1'.format(depth)],
+                                        num_filters=n_filters, filter_size=3, pad='valid',
+                                        W=HeNormal(gain='relu'),
+                                        nonlinearity=nonlinearity)
+
+    # Contraction
+    for d in range(depth):
+        #There is no pooling at the last layer
+        deepest = d == depth-1
+        contraction(d, deepest)
+
+    # Expansion
+    for d in reversed(range(depth-1)):
+        deepest = d == depth-2
+        expansion(d, deepest)
+
+    # Output layer
+    net['out'] = Conv2DLayer(net['_conv0_2'], num_filters=num_classes, filter_size=(1,1), pad='valid',
+                                    nonlinearity=None)
+
+    print ('Network output shape '+ str(lasagne.layers.get_output_shape(net['out'])))
+    return net
+
+
+def define_updates(network, input_var, target_var, weight_var, learning_rate=0.01, momentum=0.9, l2_lambda=1e-5):
+    params = lasagne.layers.get_all_params(network, trainable=True)
+
+    out = lasagne.layers.get_output(network)
+    test_out = lasagne.layers.get_output(network, deterministic=True)
+
+    l2_loss = l2_lambda * regularize_network_params(network, l2)
+
+    train_metrics = _score_metrics(out, target_var, weight_var, l2_loss)
+    loss, acc, target_prediction, prediction = train_metrics
+
+    val_metrics = _score_metrics(test_out, target_var, weight_var, l2_loss)
+    t_loss, t_acc, t_target_prediction, t_prediction = val_metrics
+
+
+    updates = lasagne.updates.nesterov_momentum(
+            loss, params, learning_rate=learning_rate, momentum=momentum)
+
+    train_fn = theano.function([input_var, target_var, weight_var],[
+                                loss, l2_loss, acc, target_prediction, prediction],
+                                updates=updates)
+
+    val_fn = theano.function([input_var, target_var, weight_var], [
+                                t_loss, l2_loss, t_acc, t_target_prediction, t_prediction])
+
+
+    return train_fn, val_fn
```

Comment:
If it is called `define_updates`, it should just return the updates. I can see that you're doing this to share some common code between training and validation, but this already fixes a lot of things (e.g., the kind of optimization to use, that you include L2 costs in validation). Maybe it's better to just expose the `_score_metrics()` under a suitable name, like `pixelwise_categorical_crossentropy()`, without the `l2` argument?